hypertrace-graphql-entity-schema/src/main/java/org/hypertrace/graphql/entity/request/EdgeRequestBuilder.java

```diff
@@ -89,31 +90,26 @@ private Single<EdgeSetGroupRequest> buildEdgeRequest(
       Stream<SelectedField> edgeSetFields,
       EdgeType edgeType) {
     Set<SelectedField> edgeFields = edgeSetFields.collect(Collectors.toUnmodifiableSet());
-    List<FilterArgument> filterArguments = this.getFilters(edgeFields);
-
-    if (!filterArguments.isEmpty() && edgeFields.size() > 1) {
-      throw Status.UNIMPLEMENTED
-          .withDescription("Cannot specify more than one edge type with edge filters")
-          .asRuntimeException();
-    }
-
-    Map<String, Set<SelectedField>> edgesByType = this.getEdgesByType(edgeFields.stream());
-    Set<SelectedField> allEdges =
-        edgesByType.values().stream()
+    Map<String, Set<SelectedField>> edgesSelectionsByType =
+        this.getEdgesSelectionByType(edgeFields.stream());
+    Set<SelectedField> allEdgesSelections =
+        edgesSelectionsByType.values().stream()
             .flatMap(Collection::stream)
             .collect(Collectors.toUnmodifiableSet());
 
+    Map<String, Set<FilterArgument>> edgesFiltersByType =
+        getEdgesFiltersByType(edgeFields.stream());
+
     return zip(
-        this.getRequestedAndRequiredAttributes(context, allEdges, edgeType),
+        this.getRequestedAndRequiredAttributes(context, allEdgesSelections, edgeType),
```

Comment:
Discussed offline. Ideally we make the abstract request changes consistent to reflect the schema - specifically that not just the filters but the selections too are broken down by type. I'd do that by removing the 4 collections (types, attributes, metrics, filters) and converting them to a single map of type -> {attributes, metrics, filters} . In the current approach, we create such a map only for filters which cause implicit assumptions (type collection should be in sync with keyset of filters) and duplication.

The fact that gateway doesn't yet support heterogeneous selections like that should be isolated to the gateway conversion. If fixing this is a dramatic scope increase we can defer it - but that does mean we're going to have to break the abstraction twice rather than just once so should be considered a last resort.